src/port/wolfssl_io.c

```diff
@@ -17,55 +16,106 @@
  * You should have received a copy of the GNU General Public License
  * along with this program; if not, write to the Free Software
  * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1335, USA
+ *
+ * wolfIP <-> wolfSSL glue for custom IO callbacks.
  */
 #include "wolfip.h"
 #include <wolfssl/wolfcrypt/settings.h>
 #include <wolfssl/ssl.h>
+#include <wolfssl/wolfcrypt/memory.h>
+
+#define MAX_WOLFIP_CTX 8
+
+struct ctx_entry {
+    WOLFSSL_CTX *ctx;
+    struct wolfIP *stack;
+};
+
+struct wolfip_io_desc {
+    int fd;
+    struct wolfIP *stack;
+};
 
-static struct wolfIP *ref_ipstack = NULL; 
+static struct ctx_entry ctx_map[MAX_WOLFIP_CTX];
+static struct wolfip_io_desc io_descs[MAX_WOLFIP_CTX];
+
+static struct wolfIP *wolfIP_lookup_stack(WOLFSSL_CTX *ctx)
+{
+    for (int i = 0; i < MAX_WOLFIP_CTX; i++) {
+        if (ctx_map[i].ctx == ctx)
+            return ctx_map[i].stack;
+    }
+    return NULL;
+}
+
+static void wolfIP_register_stack(WOLFSSL_CTX *ctx, struct wolfIP *stack)
+{
+    for (int i = 0; i < MAX_WOLFIP_CTX; i++) {
+        if (ctx_map[i].ctx == ctx || ctx_map[i].ctx == NULL) {
+            ctx_map[i].ctx = ctx;
+            ctx_map[i].stack = stack;
+            return;
+        }
+    }
+}
 
 static int wolfIP_io_recv(WOLFSSL* ssl, char* buf, int sz, void* ctx)
 {
-    int ret = 0;
-    int fd = (intptr_t)ctx;
+    struct wolfip_io_desc *desc = (struct wolfip_io_desc *)ctx;
     (void)ssl;
-    if (!ref_ipstack)
-        return -1;
-    ret = wolfIP_sock_recv(ref_ipstack, fd, buf, sz, 0);
-    if (ret == -11)
+
+    if (!desc || !desc->stack)
+        return WOLFSSL_CBIO_ERR_GENERAL;
+
+    int ret = wolfIP_sock_recv(desc->stack, desc->fd, buf, sz, 0);
+    if (ret == -11 || ret == -1)
```

Comment:
Add defines for these. API shouldn't return magic numbers.

Reply:
it is "-EINVAL" now, should be fixed. This will be compatible with posix `errno.h` inclusion. "-1" is just the callback generic error, does not have any identifier.